### PR TITLE
TSK-1538: Added transfer method with control setTransferFlag

### PR DESCRIPTION
**Pull request by tge20**
---
 
### For the submitter:
- [ ] I updated the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) and will supply links to the specific files
- [x] I did not update the [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview)
- [x] I included a link to the [SonarCloud branch analysis](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1019969636/SonarCloud+Integration)
- [x] I added a description of changes on the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [ ] I did not update the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana)
- [x] I put the ticket in review
- [ ] After integration of the pull request, I verified our [bluemix test environment](http://taskana.mybluemix.net/taskana) is not broken

### Verified by the reviewer:
- [ ] Commit message format → TSK-XXX: Your commit message.
- [ ] Submitter's update to [documentation](https://taskana.atlassian.net/wiki/spaces/TAS/overview) is sufficient
- [ ] SonarCloud analysis meets our standards
- [ ] Update of the [current release notes](https://taskana.atlassian.net/wiki/spaces/TAS/pages/1281392672/Current+Release+Notes+Taskana) reflects changes
- [ ] PR fulfills the ticket
- [ ] Edge cases and unwanted side effects are tested
- [ ] Readability

### CODE REVIEW

**Comment by tge20** [comment]
https://sonarcloud.io/dashboard?branch=TSK-1538&id=tge20_taskana

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 1]
```diff
@@ -160,39 +160,86 @@ Task createTask(Task taskToCreate)
...
+   * Transfer a Task to another Workbasket. <br>
+   * The transfer sets the transferred flag and resets the read flag.
    *
-   * @param taskId The id of the {@link Task} to be transferred
-   * @param workbasketKey The key of the target work basket
-   * @param domain The domain of the target work basket
-   * @return the transferred task
+   * @param taskId the id of the {@link Task} to be transferred
+   * @param workbasketKey the key of the target Workbasket
+   * @param domain the domain of the target Workbasket
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:209]: We should discuss this. Everywhere else we start the Javadoc with a capital letter.

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 2]
```diff
@@ -160,39 +160,86 @@ Task createTask(Task taskToCreate)
...
    * @throws TaskNotFoundException Thrown if the {@link Task} with taskId was not found.
-   * @throws WorkbasketNotFoundException Thrown if the target work basket was not found.
+   * @throws WorkbasketNotFoundException Thrown if the target Workbasket was not found.
    * @throws NotAuthorizedException Thrown if the current user is not authorized to transfer this
-   *     {@link Task} to the target work basket
-   * @throws InvalidStateException Thrown if the task is in a state which does not allow
+   *     {@link Task} to the target Workbasket
+   * @throws InvalidStateException Thrown if the Task is in a state which does not allow
    *     transferring
    */
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:217]: Same thing as above: Replace the javadoc with a `@see` reference?

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 3]
```diff
@@ -282,36 +329,84 @@ Task updateTask(Task task)
...
    *
-   * @param destinationWorkbasketId target workbasket id
-   * @param taskIds source task which will be moved
+   * @param destinationWorkbasketId target Workbasket id
+   * @param taskIds source Tasks which will be moved
    * @return Bulkresult with ID and Error in it for failed transactions.
    * @throws NotAuthorizedException if the caller hasn´t permissions on tarket WB.
    * @throws InvalidArgumentException if the method paramesters are EMPTY or NULL.
    * @throws WorkbasketNotFoundException if the target WB can´t be found.
    */
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:343]: `@see` ?

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 4]
```diff
@@ -282,36 +329,84 @@ Task updateTask(Task task)
...
+   * stored and returned in the end.
+   *
+   * @param destinationWorkbasketKey target Workbasket key
+   * @param destinationWorkbasketDomain target Workbasket domain
+   * @param taskIds source Tasks which will be moved
+   * @return Bulkresult with ID and Error in it for failed transactions.
+   * @throws NotAuthorizedException if the caller hasn´t permissions on tarket WB.
+   * @throws InvalidArgumentException if the method paramesters are EMPTY or NULL.
+   * @throws WorkbasketNotFoundException if the target WB can´t be found.
+   */
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:382]: `@see`?

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java [hunk 5]
```diff
@@ -88,9 +89,10 @@ Task transfer(String taskId, String destinationWorkbasketKey, String domain)
       Workbasket destinationWorkbasket =
           workbasketService.getWorkbasket(destinationWorkbasketKey, domain);
 
-      // reset read flag and set transferred flag
+      // reset read flag
+      // set transferred flag only if wanted
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java:93]: I don't think this comment is necessary.

File: lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java [hunk 6]
```diff
@@ -146,9 +148,10 @@ Task transfer(String taskId, String destinationWorkbasketId)
 
       Workbasket destinationWorkbasket = workbasketService.getWorkbasket(destinationWorkbasketId);
 
-      // reset read flag and set transferred flag
+      // reset read flag
+      // set transferred flag only if wanted
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/internal/TaskTransferrer.java:152]: I don't think this comment is necessary.

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java [hunk 7]
```diff
@@ -397,4 +398,46 @@ void should_BulkTransferTasks_When_WorkbasketKeyAndDomainIsProvided() throws Exc
     assertThat(transferredTask.getModified().isBefore(before)).isFalse();
     assertThat(transferredTask.getOwner()).isNull();
   }
+
+  @WithAccessId(user = "admin")
+  @Test
+  void should_NotSetTheTransferFlag_When_SetTransferFlagDisabled() throws Exception {
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:404]:
```suggestion
  void should_NotSetTheTransferFlag_When_SetTransferFlagIsNotRequested() throws Exception {
```

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java [hunk 8]
```diff
@@ -397,4 +398,46 @@ void should_BulkTransferTasks_When_WorkbasketKeyAndDomainIsProvided() throws Exc
...
+        false);
+
+    Task transferredTask = taskService.getTask("TKI:000000000000000000000000000000000003");
+    assertThat(transferredTask).isNotNull();
+    assertThat(transferredTask.isTransferred()).isFalse();
+  }
+
+  @WithAccessId(user = "admin")
+  @Test
+  void should_NotSetTheTransferFlagWithinBulkTransfer_When_SetTransferFlagDisabled()
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:418]:
```suggestion
  void should_NotSetTheTransferFlagWithinBulkTransfer_When_SetTransferFlagIsNotRequested()
```

File: lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java [hunk 9]
```diff
@@ -397,4 +398,46 @@ void should_BulkTransferTasks_When_WorkbasketKeyAndDomainIsProvided() throws Exc
...
+            .idIn(
+                "TKI:000000000000000000000000000000000004",
+                "TKI:000000000000000000000000000000000005",
+                "TKI:000000000000000000000000000000000003")
+            .list();
+
+    assertThat(transferredTasks)
+        .extracting(TaskSummary::isTransferred)
+        .contains(false)
+        .doesNotContain(true);
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:441]:
```suggestion
        .containsOnly(false);
```

Comment (mustaphazorgati) [reply lib/taskana-core/src/test/java/acceptance/task/TransferTaskAccTest.java:441]: https://joel-costigliola.github.io/assertj/core/api/org/assertj/core/api/ListAssert.html#containsOnly(ELEMENT...)

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java [hunk 10]
```diff
@@ -405,11 +406,15 @@
   @PostMapping(path = RestEndpoints.URL_TASKS_ID_TRANSFER_WORKBASKET_ID)
   @Transactional(rollbackFor = Exception.class)
   public ResponseEntity<TaskRepresentationModel> transferTask(
-      @PathVariable String taskId, @PathVariable String workbasketId)
+      @PathVariable String taskId,
+      @PathVariable String workbasketId,
+      @RequestBody(required = false) Boolean setTransferFlag)
```
Comment (mustaphazorgati) [inline rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:411]: did you test this manually with postman?

Comment (mustaphazorgati) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:411]: @gitgoodjhe do you think we should create a test for this? I am not sure..

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java (hunk 1 above)
Comment (tge20) [reply lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:209]: Yes, we will discuss it in our COP.

File: rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java (hunk 10 above)
Comment (tge20) [reply rest/taskana-rest-spring/src/main/java/pro/taskana/task/rest/TaskController.java:411]: No, unfortunately not. I'll do it now.

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 11]
```diff
@@ -160,39 +160,83 @@ Task createTask(Task taskToCreate)
...
+
+  /**
+   * Transfers a {@linkplain Task} to another {@linkplain
+   * pro.taskana.workbasket.api.models.Workbasket Workbasket}.
+   *
+   * <p>The transfer resets the read flag and sets the transfer flag if {@code setTransferFlag} is
+   * {@code true}.
+   *
+   * @param taskId the id of the {@linkplain Task} to be transferred
+   * @param destinationWorkbasketId the Id of the target {@linkplain
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:184]:
```suggestion
   * @param destinationWorkbasketId the id of the target {@linkplain
```

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 12]
```diff
@@ -160,39 +160,83 @@ Task createTask(Task taskToCreate)
...
+  }
+
+  /**
+   * Transfers a {@linkplain Task} to another {@linkplain
+   * pro.taskana.workbasket.api.models.Workbasket Workbasket}.
+   *
+   * <p>The transfer resets the read flag and sets the transfer flag if {@code setTransferFlag} is
+   * {@code true}.
+   *
+   * @param taskId the id of the {@linkplain Task} to be transferred
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:183]:
```suggestion
   * @param taskId the id of the {@linkplain Task} which should be transferred
```

File: lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java [hunk 13]
```diff
@@ -160,39 +160,83 @@ Task createTask(Task taskToCreate)
...
+   * pro.taskana.workbasket.api.models.Workbasket Workbasket}.
+   *
+   * <p>The transfer resets the read flag and sets the transfer flag if {@code setTransferFlag} is
+   * {@code true}.
    *
-   * @param taskId The id of the {@link Task} to be transferred
-   * @param workbasketKey The key of the target work basket
-   * @param domain The domain of the target work basket
-   * @return the transferred task
+   * @param taskId the id of the {@linkplain Task} to be transferred
```
Comment (mustaphazorgati) [inline lib/taskana-core/src/main/java/pro/taskana/task/api/TaskService.java:223]:
```suggestion
   * @param taskId the id of the {@linkplain Task} which should be transferred
```

**Review by mustaphazorgati — CHANGES_REQUESTED** [review]

**Pull request merged**